File: coordinator/api/tool_constraint_schema.go
package api

import (
	"encoding/json"
	"fmt"
	"math"
	"strings"
)

const (
	constrainedMaxArrayItems        = 16
	constrainedMaxGrammarComplexity = 50_000
	constrainedSchemaFixedCost      = 8
	constrainedNullableBranchCost   = 4
)

var constrainedStringDelimiters = []string{
	`<|"|>`,
	"<escape>",
	"<|tool_call>",
	"<tool_call|>",
	"<start_function_call>",
	"<end_function_call>",
}

func validateConstrainedSchema(raw any, root bool, depth int, path string) error {
	if depth > 16 {
		return unsupportedToolConstraint(path + " exceeds maximum nesting depth 16")
	}
	schema, ok := raw.(map[string]any)
	if !ok {
		return invalidToolConstraint(path+" must be a schema object", "tools")
	}
	// Normalization rewrites the allow-all `{}` / `true` schemas into a
	// render-safe marker shape that grammar modes compile as the free string
	// the original `{}` compiled to. Only that exact non-root shape is
	// accepted; any other marker-bearing schema fails closed. Mirrors the
	// Swift compiler and the Rust sidecar validator. (Production validates
	// the pre-normalization body, where the marker never legitimately
	// appears; this keeps re-validation of normalized bodies symmetric.)
	if marker, present := schema[originalBooleanSchemaKey]; present {
		if !root && marker == true && len(schema) == 2 && schema["type"] == "string" {
			return nil
		}
		return unsupportedToolConstraint(path + " uses reserved schema metadata")
	}
	for key := range schema {
		switch key {
		case "type", "properties", "required", "additionalProperties", "items",
			"minItems", "maxItems", "nullable", "enum", "const", "description",
			"title", "default", "examples", "deprecated", "readOnly", "writeOnly":
		default:
			return unsupportedToolConstraint(path + " uses " + key)
		}
	}
	kind, nullable, err := constrainedSchemaType(schema)
	if err != nil {
		return unsupportedToolConstraint(path + ": " + err.Error())
	}
	if rawNullable, exists := schema["nullable"]; exists {
		flag, ok := rawNullable.(bool)
		if !ok {
			return invalidToolConstraint(path+".nullable must be boolean", "tools")
		}
		nullable = nullable || flag
	}
	if root && (kind != "object" || nullable) {
		return invalidToolConstraint(path+" must be a non-null object schema", "tools")
	}
	if err := validateConstrainedFiniteValues(schema, kind, nullable, path); err != nil {
		return err
	}
	switch kind {
	case "object":
		properties := map[string]any{}
		if rawProperties, exists := schema["properties"]; exists {
			var ok bool
			properties, ok = rawProperties.(map[string]any)
			if !ok {
				return invalidToolConstraint(path+".properties must be an object", "tools")
			}
		}
		if len(properties) > 128 {
			return invalidToolConstraint(path+" has more than 128 properties", "tools")
		}
		for name, child := range properties {
			if !toolFunctionNamePattern.MatchString(name) {
				return unsupportedToolConstraint(
					path + " property names must match ^[a-zA-Z0-9_-]{1,64}$")
			}
			if err := validateConstrainedSchema(child, false, depth+1, path+".properties."+name); err != nil {
				return err
			}
		}
		if required, exists := schema["required"]; exists {
			values, ok := required.([]any)
			if !ok {
				return invalidToolConstraint(path+".required must be an array", "tools")
			}
			seen := make(map[string]struct{}, len(values))
			for _, value := range values {
				name, ok := value.(string)
				if !ok {
					return invalidToolConstraint(path+".required members must be strings", "tools")
				}
				if _, declared := properties[name]; !declared {
					return invalidToolConstraint(path+".required contains an undeclared property", "tools")
				}
				if _, duplicate := seen[name]; duplicate {
					return invalidToolConstraint(path+".required contains a duplicate property", "tools")
				}
				seen[name] = struct{}{}
			}
		}
		if additional, exists := schema["additionalProperties"]; exists {
			if _, ok := additional.(bool); !ok {
				return unsupportedToolConstraint(path + ".additionalProperties must be boolean")
			}
		}
	case "array":
		items, exists := schema["items"]
		if !exists {
			return unsupportedToolConstraint(path + " array requires a single items schema")
		}
		if err := validateConstrainedSchema(items, false, depth+1, path+".items"); err != nil {
			return err
		}
		rawMinItems, hasMinItems := schema["minItems"]
		if hasMinItems && rawMinItems == nil {
			return invalidToolConstraint(path+".minItems must be a nonnegative integer", "tools")
		}
		minItems, err := constrainedNonnegativeInt(rawMinItems, 0)
		if err != nil {
			return invalidToolConstraint(path+".minItems must be a nonnegative integer", "tools")
		}
		rawMaxItems, hasMaxItems := schema["maxItems"]
		if hasMaxItems && rawMaxItems == nil {
			return invalidToolConstraint(path+".maxItems must be a nonnegative integer", "tools")
		}
		maxItems, hasMax, err := constrainedOptionalInt(rawMaxItems)
		if err != nil {
			return invalidToolConstraint(path+".maxItems must be a nonnegative integer", "tools")
		}
		if minItems > constrainedMaxArrayItems ||
			(hasMax && (maxItems > constrainedMaxArrayItems || maxItems < minItems)) {
			return unsupportedToolConstraint(path + " array bounds must be within 0...16")
		}
	case "string", "boolean", "integer", "number", "null":
	default:
		return unsupportedToolConstraint(path + " has unsupported type " + kind)
	}
	return nil
}

func constrainedSchemaGrammarCost(raw any) int {
	schema, ok := raw.(map[string]any)
	if !ok {
		return constrainedMaxGrammarComplexity + 1
	}
	kind, nullable, err := constrainedSchemaType(schema)
	if err != nil {
		return constrainedMaxGrammarComplexity + 1
	}
	if rawNullable, exists := schema["nullable"]; exists {
		flag, ok := rawNullable.(bool)
		if !ok {
			return constrainedMaxGrammarComplexity + 1
		}
		nullable = nullable || flag
	}
	values, finite := constrainedSchemaFiniteValues(schema)
	// JSON Schema applies type and enum/const conjunctively: a nullable type
	// admits null only when the finite value set itself contains null, so the
	// grammar builds no null branch (and charges no branch cost) otherwise.
	// Mirrors the Swift compiler's effective-nullable rule.
	if finite && !constrainedValuesContainNull(values) {
		nullable = false
	}
	baseCost := constrainedSchemaFixedCost
	if nullable {
		baseCost = constrainedGrammarAdd(baseCost, constrainedNullableBranchCost)
	}
	payloadCost := 0
	switch kind {
	case "object":
		payloadCost = 2
		properties, _ := schema["properties"].(map[string]any)
		for name, child := range properties {
			payloadCost = constrainedGrammarAdd(
				payloadCost, len([]byte(name))+2+constrainedSchemaGrammarCost(child))
		}
	case "array":
		count := constrainedMaxArrayItems
		if maximum, present, parseErr := constrainedOptionalInt(schema["maxItems"]); parseErr == nil && present {
			count = maximum
		}
		itemCost := constrainedSchemaGrammarCost(schema["items"])
		payloadCost = constrainedGrammarAdd(
			2, constrainedGrammarMultiply(itemCost+1, count))
	case "string":
		if !finite {
			payloadCost = 16
		} else {
			for _, value := range values {
				if text, ok := value.(string); ok {
					payloadCost = constrainedGrammarAdd(
						payloadCost, len([]byte(text))+10)
				}
			}
		}
	case "boolean":
		if !finite {
			payloadCost = 10
		} else {
			count := 0
			for _, value := range values {
				if _, ok := value.(bool); ok {
					count++
				}
			}
			payloadCost = constrainedGrammarMultiply(count, 5)
		}
	case "integer", "number":
		if !finite {
			if kind == "integer" {
				payloadCost = 20
			} else {
				payloadCost = 40
			}
		} else {
			for _, value := range values {
				if number, ok := value.(json.Number); ok {
					payloadCost = constrainedGrammarAdd(
						payloadCost, len(number.String()))
				}
			}
		}
	case "null":
		payloadCost = 4
	default:
		return constrainedMaxGrammarComplexity + 1
	}
	return constrainedGrammarAdd(baseCost, payloadCost)
}

func constrainedSchemaFiniteValues(schema map[string]any) ([]any, bool) {
	if constant, ok := schema["const"]; ok {
		return []any{constant}, true
	}
	values, ok := schema["enum"].([]any)
	return values, ok
}

func constrainedValuesContainNull(values []any) bool {
	for _, value := range values {
		if value == nil {
			return true
		}
	}
	return false
}

func constrainedGrammarAdd(lhs, rhs int) int {
	if lhs > constrainedMaxGrammarComplexity ||
		rhs > constrainedMaxGrammarComplexity ||
		lhs > constrainedMaxGrammarComplexity-rhs {
		return constrainedMaxGrammarComplexity + 1
	}
	return lhs + rhs
}

func constrainedGrammarMultiply(lhs, rhs int) int {
	if lhs < 0 || rhs < 0 ||
		(rhs != 0 && lhs > constrainedMaxGrammarComplexity/rhs) {
		return constrainedMaxGrammarComplexity + 1
	}
	product := lhs * rhs
	if product > constrainedMaxGrammarComplexity {
		return constrainedMaxGrammarComplexity + 1
	}
	return product
}

func validateConstrainedFiniteValues(
	schema map[string]any,
	kind string,
	nullable bool,
	path string,
) error {
	constant, hasConstant := schema["const"]
	rawEnum, hasEnum := schema["enum"]
	if hasConstant && hasEnum {
		return invalidToolConstraint(path+" cannot contain both const and enum", "tools")
	}
	var values []any
	if hasConstant {
		values = []any{constant}
	} else if hasEnum {
		var ok bool
		values, ok = rawEnum.([]any)
		if !ok || len(values) == 0 || len(values) > 128 {
			return invalidToolConstraint(path+".enum must contain 1...128 values", "tools")
		}
	} else {
		return nil
	}
	if kind == "object" || kind == "array" || kind == "number" {
		return unsupportedToolConstraint(path + " uses enum/const on " + kind)
	}
	for _, value := range values {
		if value == nil && nullable {
			continue
		}
		if kind == "string" {
			if text, ok := value.(string); ok {
				for _, marker := range constrainedStringDelimiters {
					if strings.Contains(text, marker) {
						return unsupportedToolConstraint(
							path + " string enum/const contains a Gemma parser delimiter")
					}
				}
			}
		}
		if !constrainedValueMatchesType(value, kind) {
			return invalidToolConstraint(
				path+" enum/const does not match type "+kind, "tools")
		}
	}
	return nil
}

func constrainedValueMatchesType(value any, kind string) bool {
	switch kind {
	case "string":
		_, ok := value.(string)
		return ok
	case "boolean":
		_, ok := value.(bool)
		return ok
	case "integer":
		number, ok := value.(json.Number)
		if !ok {
			return false
		}
		_, err := number.Int64()
		return err == nil
	case "number":
		number, ok := value.(json.Number)
		if !ok {
			return false
		}
		parsed, err := number.Float64()
		return err == nil &&
			!math.IsInf(parsed, 0) &&
			!math.IsNaN(parsed)
	case "null":
		return value == nil
	default:
		return false
	}
}

func constrainedSchemaType(schema map[string]any) (string, bool, error) {
	raw := schema["type"]
	if raw == nil {
		if schema["properties"] != nil || schema["additionalProperties"] != nil {
			return "object", false, nil
		}
		if schema["items"] != nil {
			return "array", false, nil
		}
		return "string", false, nil
	}
	if value, ok := raw.(string); ok {
		return strings.ToLower(value), false, nil
	}
	values, ok := raw.([]any)
	if !ok {
		return "", false, fmt.Errorf("type must be a string")
	}
	var nonNull string
	sawNull := false
	for _, rawValue := range values {
		value, ok := rawValue.(string)
		if !ok {
			return "", false, fmt.Errorf("type members must be strings")
		}
		if strings.EqualFold(value, "null") {
			sawNull = true
		} else if nonNull == "" {
			nonNull = strings.ToLower(value)
		} else {
			return "", false, fmt.Errorf("only one type plus null is supported")
		}
	}
	if !sawNull || nonNull == "" {
		return "", false, fmt.Errorf("only one type plus null is supported")
	}
	return nonNull, true, nil
}

func constrainedNonnegativeInt(raw any, fallback int) (int, error) {
	if raw == nil {
		return fallback, nil
	}
	number, ok := raw.(json.Number)
	if !ok {
		return 0, fmt.Errorf("not an integer")
	}
	value, err := number.Int64()
	if err != nil || value < 0 || value > int64(^uint(0)>>1) {
		return 0, fmt.Errorf("not a nonnegative integer")
	}
	return int(value), nil
}

func constrainedOptionalInt(raw any) (int, bool, error) {
	if raw == nil {
		return 0, false, nil
	}
	value, err := constrainedNonnegativeInt(raw, 0)
	return value, true, err
}

File: provider-swift/Sources/InferenceWorkerCore/Runtime/Inference/ToolConstraintSchema.swift
// Copyright © 2026 Eigen Labs.
//
// Fail-closed JSON-Schema subset used by forced tool-choice enforcement.
// Gemma consumes it in the sampler grammar; Qwen consumes it at the withheld
// parser/validation boundary before any call is exposed.
// The subset is intentionally small and exact. Unsupported assertion keywords
// are rejected for required/named choices instead of being prompt-only theater.

import Foundation
import MLXLMCommon
import MLXLMServer

private typealias SchemaJSONValue = MLXLMCommon.JSONValue

enum ToolConstraintMode: Sendable, Equatable {
    case none
    case auto
    case required
    case named(String)

    var requiresInferenceConstraint: Bool {
        switch self {
        case .none, .required, .named: true
        case .auto: false
        }
    }

    var telemetryValue: String {
        switch self {
        case .none: "none"
        case .auto: "auto"
        case .required: "required"
        case .named: "named"
        }
    }
}

struct CompiledToolSchema: Sendable {
    let name: String
    let parameters: ToolValueSchema
}

indirect enum ToolValueSchema: Sendable {
    struct Property: Sendable {
        let name: String
        let schema: ToolValueSchema
        let required: Bool
    }

    case object(properties: [Property], allowsAdditional: Bool, nullable: Bool)
    case array(items: ToolValueSchema, minItems: Int, maxItems: Int?, nullable: Bool)
    case string(values: [String]?, nullable: Bool)
    case boolean(values: [Bool]?, nullable: Bool)
    case integer(values: [Int]?, nullable: Bool)
    case number(values: [Double]?, nullable: Bool)
    case null

    var nullable: Bool {
        switch self {
        case .object(_, _, let nullable),
            .array(_, _, _, let nullable),
            .string(_, let nullable),
            .boolean(_, let nullable),
            .integer(_, let nullable),
            .number(_, let nullable):
            nullable
        case .null:
            true
        }
    }
}

enum ToolConstraintSchemaError: LocalizedError, Equatable {
    case invalid(String)
    case unsupported(String)

    var errorDescription: String? {
        switch self {
        case .invalid(let detail):
            "invalid tool constraint schema: \(detail)"
        case .unsupported(let detail):
            "unsupported tool constraint schema: \(detail)"
        }
    }
}

enum ToolConstraintSchemaCompiler {
    static let maxTools = 64
    static let maxProperties = 128
    static let maxDepth = 16
    static let maxArrayItems = 16
    static let maxGrammarComplexity = 50_000

    static func compile(
        tools: [OpenAITool]?,
        mode: ToolConstraintMode
    ) throws -> [CompiledToolSchema] {
        let tools = tools ?? []
        guard tools.count <= maxTools else {
            throw ToolConstraintSchemaError.invalid("at most \(maxTools) tools are allowed")
        }
        if mode.requiresInferenceConstraint {
            switch mode {
            case .required, .named:
                guard !tools.isEmpty else {
                    throw ToolConstraintSchemaError.invalid(
                        "required tool_choice needs at least one declared function")
                }
            case .none, .auto:
                break
            }
        }

        var names = Set<String>()
        var compiled: [CompiledToolSchema] = []
        var grammarComplexity = 0
        compiled.reserveCapacity(tools.count)
        for tool in tools {
            guard tool.type == "function" else {
                throw ToolConstraintSchemaError.unsupported(
                    "only function tools can be inference-constrained")
            }
            let name = tool.function.name
            guard ToolChoicePromptPolicy.isValidFunctionName(name) else {
                throw ToolConstraintSchemaError.invalid(
                    "tool function names must match ^[a-zA-Z0-9_-]{1,64}$")
            }
            guard names.insert(name).inserted else {
                throw ToolConstraintSchemaError.invalid(
                    "tool function names must be unique")
            }
            let raw = tool.function.parameters ?? .object(["type": .string("object")])
            let schema = try compileSchema(
                raw, depth: 0, path: "\(name).parameters",
                allowRenderMetadata: mode.requiresInferenceConstraint)
            guard case .object = schema else {
                throw ToolConstraintSchemaError.invalid(
                    "\(name).parameters must have type object")
            }
            grammarComplexity += name.utf8.count + grammarCost(schema)
            guard grammarComplexity <= maxGrammarComplexity else {
                throw ToolConstraintSchemaError.unsupported(
                    "combined tool grammar exceeds the \(maxGrammarComplexity)-unit safety limit")
            }
            compiled.append(CompiledToolSchema(name: name, parameters: schema))
        }

        if case .named(let selected) = mode, !names.contains(selected) {
            throw ToolConstraintSchemaError.invalid(
                "tool_choice names an undeclared function")
        }
        return compiled
    }

    private static func grammarCost(_ schema: ToolValueSchema) -> Int {
        let nullableBranchCost: Int
        if case .null = schema {
            nullableBranchCost = 0
        } else {
            nullableBranchCost = schema.nullable ? 4 : 0
        }
        let baseCost = boundedGrammarAdd(
            8, nullableBranchCost)
        let payloadCost: Int
        switch schema {
        case .object(let properties, _, _):
            payloadCost = properties.reduce(2) {
                boundedGrammarAdd(
                    $0,
                    boundedGrammarAdd(
                        $1.name.utf8.count + 2,
                        grammarCost($1.schema)))
            }
        case .array(let items, _, let maxItems, _):
            let count = maxItems ?? maxArrayItems
            payloadCost = boundedGrammarAdd(
                2,
                boundedGrammarMultiply(
                    boundedGrammarAdd(grammarCost(items), 1),
                    count))
        case .string(let values, _):
            payloadCost =
                values?.reduce(0) {
                    boundedGrammarAdd($0, $1.utf8.count + 10)
                } ?? 16
        case .boolean(let values, _):
            payloadCost = boundedGrammarMultiply(values?.count ?? 2, 5)
        case .integer(let values, _):
            payloadCost =
                values?.reduce(0) {
                    boundedGrammarAdd($0, String($1).count)
                } ?? 20
        case .number(let values, _):
            payloadCost =
                values?.reduce(0) {
                    boundedGrammarAdd($0, String($1).count)
                } ?? 40
        case .null:
            payloadCost = 4
        }
        return boundedGrammarAdd(baseCost, payloadCost)
    }

    private static func boundedGrammarAdd(_ lhs: Int, _ rhs: Int) -> Int {
        let (sum, overflow) = lhs.addingReportingOverflow(rhs)
        return overflow ? maxGrammarComplexity + 1 : min(maxGrammarComplexity + 1, sum)
    }

    private static func boundedGrammarMultiply(_ lhs: Int, _ rhs: Int) -> Int {
        let (product, overflow) = lhs.multipliedReportingOverflow(by: rhs)
        return overflow
            ? maxGrammarComplexity + 1
            : min(maxGrammarComplexity + 1, product)
    }

    private static func compileSchema(
        _ raw: SchemaJSONValue,
        depth: Int,
        path: String,
        allowRenderMetadata: Bool
    ) throws -> ToolValueSchema {
        guard depth <= maxDepth else {
            throw ToolConstraintSchemaError.unsupported(
                "\(path) exceeds maximum nesting depth \(maxDepth)")
        }
        guard case .object(let object) = raw else {
            throw ToolConstraintSchemaError.invalid("\(path) must be a schema object")
        }

        // Normalization rewrites the allow-all `{}` / `true` schemas into a
        // render-safe marker shape. In grammar modes (required/named) that
        // shape compiles to the free string the original `{}` compiled to
        // before normalization carried the marker. In auto mode the marker
        // stays unsupported so compilation falls back to raw validation,
        // which restores the exact allow-all semantics — a compiled string
        // grammar would wrongly reject schema-valid non-string values.
        if let marker = object[ToolSchemaNormalization.originalBooleanSchemaKey] {
            guard allowRenderMetadata,
                marker == .bool(true),
                object.count == 2,
                object["type"] == .string("string")
            else {
                throw ToolConstraintSchemaError.unsupported(
                    "\(path) uses reserved schema metadata")
            }
            return .string(values: nil, nullable: false)
        }

        // A DIRECT (un-normalized) empty `{}` schema is the same allow-all
        // case: grammar modes keep compiling it as a free string, but auto
        // must fall back to raw validation instead of letting the inferred
        // string type reject schema-valid non-string values.
        if !allowRenderMetadata, object.isEmpty {
            throw ToolConstraintSchemaError.unsupported(
                "\(path) empty schemas accept any value")
        }

        let supported: Set<String> = [
            "type", "properties", "required", "additionalProperties", "items",
            "minItems", "maxItems", "nullable", "enum", "const", "description",
            "title", "default", "examples", "deprecated", "readOnly", "writeOnly",
        ]
        if let keyword = object.keys.first(where: { !supported.contains($0) }) {
            throw ToolConstraintSchemaError.unsupported(
                "\(path) uses \(keyword)")
        }

        let explicitNullable = try bool(
            object["nullable"], default: false, path: "\(path).nullable")
        let (kind, typeNullable) = try schemaType(
            object, path: "\(path).type")
        var nullable = explicitNullable || typeNullable
        let enumValues = try finiteValues(object: object, path: path)
        // JSON Schema applies `type` and `enum`/`const` conjunctively: a
        // nullable type admits null only when the finite value set itself
        // contains null. Without this the grammar would build a null branch —
        // and the post-validator would accept null — for values the declared
        // enum excludes (e.g. `{"type":["string","null"],"enum":["ok"]}`).
        if let enumValues, !enumValues.contains(.null) {
            nullable = false
        }

        switch kind {
        case "object":
            return try compileObject(
                object, finiteValues: enumValues, nullable: nullable,
                depth: depth, path: path,
                allowRenderMetadata: allowRenderMetadata)
        case "array":
            guard enumValues == nil else {
                throw ToolConstraintSchemaError.unsupported(
                    "\(path) uses enum/const on an array")
            }
            guard let items = object["items"] else {
                throw ToolConstraintSchemaError.unsupported(
                    "\(path) array requires a single items schema")
            }
            let minItems = try nonnegativeInt(
                object["minItems"], default: 0, path: "\(path).minItems")
            let maxItems = try optionalNonnegativeInt(
                object["maxItems"], path: "\(path).maxItems")
            guard minItems <= maxArrayItems,
                maxItems.map({ $0 <= maxArrayItems && $0 >= minItems }) ?? true
            else {
                throw ToolConstraintSchemaError.unsupported(
                    "\(path) array bounds must be within 0...\(maxArrayItems)")
            }
            return .array(
                items: try compileSchema(
                    items, depth: depth + 1, path: "\(path).items",
                    allowRenderMetadata: allowRenderMetadata),
                minItems: minItems,
                maxItems: maxItems,
                nullable: nullable)
        case "string":
            try rejectKeywords(
                object, ["pattern", "format", "minLength", "maxLength"],
                path: path)
            let values = try enumValues.map { values in
                try values.compactMap { value -> String? in
                    if value == .null, nullable { return nil }
                    guard case .string(let string) = value else {
                        throw ToolConstraintSchemaError.invalid(
                            "\(path) enum/const does not match type string")
                    }
                    return string
                }
            }
            return .string(values: values, nullable: nullable)
        case "boolean":
            let values = try enumValues.map { values in
                try values.compactMap { value -> Bool? in
                    if value == .null, nullable { return nil }
                    guard case .bool(let flag) = value else {
                        throw ToolConstraintSchemaError.invalid(
                            "\(path) enum/const does not match type boolean")
                    }
                    return flag
                }
            }
            return .boolean(values: values, nullable: nullable)
        case "integer":
            try rejectNumericAssertions(object, path: path)
            let values = try enumValues.map { values in
                try values.compactMap { value -> Int? in
                    if value == .null, nullable { return nil }
                    let integer: Int?
                    switch value {
                    case .int(let value):
                        integer = value
                    case .double(let value)
                    where value.isFinite
                        && value.rounded() == value
                        && abs(value) < 9_007_199_254_740_992.0:
                        integer = Int(exactly: value)
                    default:
                        integer = nil
                    }
                    guard let integer else {
                        throw ToolConstraintSchemaError.invalid(
                            "\(path) enum/const does not match type integer")
                    }
                    return integer
                }
            }
            return .integer(values: values, nullable: nullable)
        case "number":
            try rejectNumericAssertions(object, path: path)
            guard enumValues == nil else {
                throw ToolConstraintSchemaError.unsupported(
                    "\(path) uses enum/const on number; use integer or string for exact choices")
            }
            return .number(values: nil, nullable: nullable)
        case "null":
            guard enumValues == nil || enumValues == [.null] else {
                throw ToolConstraintSchemaError.invalid(
                    "\(path) enum/const does not match type null")
            }
            return .null
        default:
            throw ToolConstraintSchemaError.unsupported(
                "\(path) has unsupported type \(kind)")
        }
    }

    private static func compileObject(
        _ object: [String: SchemaJSONValue],
        finiteValues: [SchemaJSONValue]?,
        nullable: Bool,
        depth: Int,
        path: String,
        allowRenderMetadata: Bool
    ) throws -> ToolValueSchema {
        guard finiteValues == nil else {
            throw ToolConstraintSchemaError.unsupported(
                "\(path) uses enum/const on an object")
        }
        let rawProperties: [String: SchemaJSONValue]
        if let value = object["properties"] {
            guard case .object(let properties) = value else {
                throw ToolConstraintSchemaError.invalid(
                    "\(path).properties must be an object")
            }
            rawProperties = properties
        } else {
            rawProperties = [:]
        }
        guard rawProperties.count <= maxProperties else {
            throw ToolConstraintSchemaError.invalid(
                "\(path) has more than \(maxProperties) properties")
        }
        let required: Set<String>
        if let value = object["required"] {
            guard case .array(let values) = value else {
                throw ToolConstraintSchemaError.invalid(
                    "\(path).required must be an array")
            }
            var names = Set<String>()
            for value in values {
                guard case .string(let name) = value, rawProperties[name] != nil else {
                    throw ToolConstraintSchemaError.invalid(
                        "\(path).required contains an undeclared property")
                }
                guard names.insert(name).inserted else {
                    throw ToolConstraintSchemaError.invalid(
                        "\(path).required contains a duplicate property")
                }
            }
            required = names
        } else {
            required = []
        }

        let allowsAdditional: Bool
        if let additional = object["additionalProperties"] {
            guard case .bool(let flag) = additional else {
                throw ToolConstraintSchemaError.unsupported(
                    "\(path).additionalProperties must be boolean")
            }
            allowsAdditional = flag
        } else {
            allowsAdditional = true
        }

        var properties: [ToolValueSchema.Property] = []
        properties.reserveCapacity(rawProperties.count)
        for (name, raw) in rawProperties.sorted(by: { $0.key < $1.key }) {
            guard ToolChoicePromptPolicy.isValidFunctionName(name) else {
                throw ToolConstraintSchemaError.unsupported(
                    "\(path) property names must match ^[a-zA-Z0-9_-]{1,64}$")
            }
            properties.append(.init(
                name: name,
                schema: try compileSchema(
                    raw, depth: depth + 1, path: "\(path).properties.\(name)",
                    allowRenderMetadata: allowRenderMetadata),
                required: required.contains(name)))
        }
        return .object(
            properties: properties,
            allowsAdditional: allowsAdditional,
            nullable: nullable)
    }

    private static func schemaType(
        _ object: [String: SchemaJSONValue],
        path: String
    ) throws -> (String, Bool) {
        guard let value = object["type"] else {
            if object["properties"] != nil || object["additionalProperties"] != nil {
                return ("object", false)
            }
            if object["items"] != nil {
                return ("array", false)
            }
            return ("string", false)
        }
        if value == .null {
            if object["properties"] != nil || object["additionalProperties"] != nil {
                return ("object", false)
            }
            if object["items"] != nil {
                return ("array", false)
            }
            return ("string", false)
        }
        if case .string(let type) = value { return (type.lowercased(), false) }
        if case .array(let members) = value {
            let types = try members.map { member -> String in
                guard case .string(let type) = member else {
                    throw ToolConstraintSchemaError.invalid("\(path) members must be strings")
                }
                return type.lowercased()
            }
            let nonNull = types.filter { $0 != "null" }
            guard types.contains("null"), nonNull.count == 1 else {
                throw ToolConstraintSchemaError.unsupported(
                    "\(path) supports only one type plus null")
            }
            return (nonNull[0], true)
        }
        throw ToolConstraintSchemaError.invalid("\(path) must be a string")
    }

    private static func finiteValues(
        object: [String: SchemaJSONValue],
        path: String
    ) throws -> [SchemaJSONValue]? {
        if let constant = object["const"] {
            guard object["enum"] == nil else {
                throw ToolConstraintSchemaError.invalid(
                    "\(path) cannot contain both const and enum")
            }
            return [constant]
        }
        guard let raw = object["enum"] else { return nil }
        guard case .array(let values) = raw, !values.isEmpty, values.count <= 128 else {
            throw ToolConstraintSchemaError.invalid(
                "\(path).enum must contain 1...128 values")
        }
        return values
    }

    private static func rejectKeywords(
        _ object: [String: SchemaJSONValue],
        _ keys: [String],
        path: String
    ) throws {
        if let key = keys.first(where: { object[$0] != nil }) {
            throw ToolConstraintSchemaError.unsupported("\(path) uses \(key)")
        }
    }

    private static func rejectNumericAssertions(
        _ object: [String: SchemaJSONValue],
        path: String
    ) throws {
        try rejectKeywords(
            object,
            ["minimum", "maximum", "exclusiveMinimum", "exclusiveMaximum", "multipleOf"],
            path: path)
    }

    private static func bool(
        _ value: SchemaJSONValue?,
        default defaultValue: Bool,
        path: String
    ) throws -> Bool {
        guard let value else { return defaultValue }
        guard case .bool(let flag) = value else {
            throw ToolConstraintSchemaError.invalid("\(path) must be boolean")
        }
        return flag
    }

    private static func nonnegativeInt(
        _ value: SchemaJSONValue?,
        default defaultValue: Int,
        path: String
    ) throws -> Int {
        guard let value else { return defaultValue }
        let number: Int?
        switch value {
        case .int(let integer):
            number = integer
        case .double(let decimal) where decimal.isFinite:
            number = Int(exactly: decimal)
        default:
            number = nil
        }
        guard let number, number >= 0 else {
            throw ToolConstraintSchemaError.invalid("\(path) must be a nonnegative integer")
        }
        return number
    }

    private static func optionalNonnegativeInt(
        _ value: SchemaJSONValue?,
        path: String
    ) throws -> Int? {
        guard let value else { return nil }
        return try nonnegativeInt(value, default: 0, path: path)
    }
}
